src/modules/emote_autocomplete/twitch/EmoteAutocomplete.jsx

```diff
@@ -0,0 +1,169 @@
+import twitch from '../../../utils/twitch.js';
+import watcher from '../../../watcher.js';
+import emoteMenuViewStore from '../../../common/stores/emote-menu-view-store.js';
+import dom from '../../../observers/dom.js';
+import emotes from '../../emotes/index.js';
+import {createSrcSet} from '../../../utils/image.js';
+import {SettingIds} from '../../../constants.js';
+import settings from '../../../settings.js';
+import './styles.module.css';
+
+const EMOTE_SET_CODE_PREFIX = '__bttv-';
+const AUTOCOMPLETE_MATCH_IMAGE_QUERY = '.emote-autocomplete-provider__image, .chat-line__message--emote';
+
+function seralizeCode(code) {
+  return `${EMOTE_SET_CODE_PREFIX}${code}`;
+}
+
+function deseralizeCode(code) {
+  if (!code.startsWith(EMOTE_SET_CODE_PREFIX)) {
+    return null;
+  }
+
+  return code.replace(EMOTE_SET_CODE_PREFIX, '');
+}
+
+function getAutocompleteEmoteProvider() {
+  const autocompleteNode = twitch.getAutocompleteProviders();
+  if (autocompleteNode == null) {
+    return null;
+  }
+
+  const autocompleteProviders = autocompleteNode.stateNode.providers;
+  return autocompleteProviders.find(({autocompleteType}) => autocompleteType === 'emote');
+}
+
+function createTwitchEmoteSet({category, emotes: categoryEmotes}) {
+  return {
+    emotes: categoryEmotes.map(({code}) => ({
+      id: seralizeCode(code),
+      modifiers: null,
+      setID: category.id,
+      token: code,
+    })),
+    id: category.id,
+  };
+}
+
+async function injectEmoteSets() {
+  const autocompleteEmoteProvider = getAutocompleteEmoteProvider();
+  if (autocompleteEmoteProvider == null) {
+    return;
+  }
+
+  const emoteCategories = emoteMenuViewStore.getProvidersCategories();
+  for (const category of emoteCategories) {
+    if (category.emotes.length === 0) {
+      continue;
+    }
+
+    const emoteSet = createTwitchEmoteSet(category);
+    const index = autocompleteEmoteProvider.props.emotes.findIndex(({id}) => id === category.category.id);
+
+    if (index === -1) {
+      autocompleteEmoteProvider.props.emotes.push(emoteSet);
```

Comment:
could we pull `emotes` out to a variable instead of accessing this attribute repeatedly?